module/zfs/arc.c

```diff
@@ -6431,6 +6428,17 @@ arc_read(zio_t *pio, spa_t *spa, const blkptr_t *bp,
 			}
 		}
 
+		/*
+		 * Gracefully handle a damaged logical block size as a
+		 * checksum error.
+		 */
+		if (lsize > spa_maxblocksize(spa)) {
+			rc = SET_ERROR(ECKSUM);
+			if (hash_lock != NULL)
+				mutex_exit(hash_lock);
+			goto out;
+		}
```

Comment:
We should be able to drop this hunk entirely since it'll be handled by the `zfs_blkptr_verify()` check in `arc_read()`.  We also want to remove this same check from `zio_read()` to resolve the deadlock.